### PR TITLE
S3353: Error message unprecise

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #5868

### CODE REVIEW

File: analyzers/its/expected/Net5/recordsvsclasses-{96F1111E-2713-4CDE-B2D8-BF973EB5A4DE}-net5.0-S3353.json
Comment (martin-strecker-sonarsource) [inline analyzers/its/expected/Net5/recordsvsclasses-{96F1111E-2713-4CDE-B2D8-BF973EB5A4DE}-net5.0-S3353.json:5]: This is a bit unfortunate but it compiles and is not wrong. I added a test to document it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs:471]: > A better error message would be:
Add the 'const' modifier to 'message', replace 'var' with 'string', and use string concatenation instead of an interpolated string.

I don't agree with this case. What about
> $"This is a {const1} {const2} {const3} {const4} ... {const10}";

do we really want to tell users to use string concatenation?

IMO we shouldn't raise an issue in this case, to avoid noise

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs:471]: The scope for interpolated string constants is fairly limited:

> Interpolated strings ${}, provided that all components are constant expressions of type string and all interpolated components lack alignment and format specifiers.

[Source](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/constant_interpolated_strings)

There are only two use cases for this AFAIK:

1. Building a constant string out of other constant strings
2. `nameof`

In your example, `const1` .. `const10` must all be string constants to raise the rule.
According to the spec, it is always possible to use the contact operator for any interpolated string, that can be turned into a const:

> The current + operator for string concatnation can combine strings in a similar manner to the current proposal.

Therefore the suggestion is always feasible.

I think these options are valid:

1. Raise and provide guidance on how to turn it into a valid const (current behavior)
2. Don't raise for verbatim interpolated strings for projects lower than C# 10.
3. Don't raise any interpolated string for projects lower than C#10.

What do you think?

Comment (andrei-epure-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs:471]: > and use string concatenation instead of an interpolated string.

> In your example, const1 .. const10 must all be string constants to raise the rule.

Sure, my main question is why would

> const string s3 = "This is a " + nameof(Message) + " some other text " + otherConst; // and another 4 concatenations

be more readable than using an interpolated string?

the `" + "` makes code less readable IMO

I would say not raise the issue when only concatenation is the answer to avoid subjective interpretation (so yes, don't raise when less than C# 10).

Comment (andrei-epure-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs:471]: An example of subjective interpretation: "I find string interpolation more readable than concatenation, so I don't want to use `const string` in this case".

I am open to other ideas though, if you think this won't spark subjective reactions from devs.

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs:471]: I also prefer to not raise the issue. I just wanted to make sure, you have the full context. Let me change the implementation.

Comment (andrei-epure-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs]: The context is: we shouldn't raise issues which are too controversial (noise) especially for SonarWay rules.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:38]: I suggest to move `$", and replace 'var' with` here

```suggestion
        private const string MessageFormat = "Add the 'const' modifier to '{0}', and replace 'var' with {1}.";
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:71]: can you move this check before the `GetDeclaredSymbol`, to shortcircuit the logic if needed?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:68]: can you move `c.SemanticModel.GetDeclaredSymbol(v)` inside `HasMutableUsagesInMethod` (the only place it's used) and lazy load it only if necessary (`MatchesIdentifier`)?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:140]: I'd suggest a later PR to limit the changes

good catch!

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:71]: and maybe add a short comment saying why we ignore string interpolation below C# 10?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs:470]: the message is confusing - what is only valid in C# 10 and above? it doesn't compile otherwise?

maybe mention that `constant string interpolation is only valid in C# 10 and above"

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:38]: The `, and replace 'var' with {1}` part is conditional. We can either use `{1}` for the whole part or use a second `MessageFormat`.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:140]: I made me a note for later.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:146]: The lazy loading here is questionable. The likelihood of finding an identifier within a method is close to 100%:
```cs
 return parentSyntax
                .DescendantNodes()
                .OfType<IdentifierNameSyntax>()
                .Where(MatchesIdentifier)
                .Any(x => IsMutatingUse(semanticModel, x));
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:38]: I'd prefer a second message format, otherwise it's not very clear

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:38]: With a second message, we need to rewrite `SupportedDiagnostics` too. `messageFormat` is mandatory for the DiagnosticDescriptor. Therefore we need two `Rule` properties and return only one of these in `SupportedDiagnostics`. That seems unfortunate to me, or do I miss something?

```cs
        private const string MessageFormat = "Add the 'const' modifier to '{0}'{1}.";
        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
```

Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:38]: I feel an itch that the messages are in two different parts of the code, maybe if the second optional part would be under MessageFormat my itch would go away

Or like: https://github.com/SonarSource/sonar-dotnet/blob/b4b5df2265374be44123d99afde8067ed99358f2/analyzers/src/SonarAnalyzer.CSharp/Rules/BinaryOperationWithIdenticalExpressions.cs#L35

```DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, "{0}");```

and have two message formats

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:148]: I think that `GetDeclaredSymbol` can return null, right?  should do a null check before calling Equals?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs]: `IndexerDeclarationSyntax` can safely be removed. It is covered by AccessorDeclarationSyntax already. Test case:
```cs
        public byte this[int idx] // IndexerDeclaration
        {
            get  // AccessorDeclaration
            {
                var pos = 0;
                var seen = 0;

                seen += 1;
                pos += 1;

                return 1;
            }
        }
```

**Comment by martin-strecker-sonarsource** [comment]
@andrei-epure-sonarsource I need another review because I fixed a small bug I found in 6886a6f (clean as you code).

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM, please fix misnomers before merging

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/SyntaxHelperTest.cs
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/SyntaxHelperTest.cs:106]: should the name of the test be `IsInTupleAssignmentTarget_IsTrue`?

otherwise it's really confusing

please update line 122 IsAssignmentToTuple_IsFalse as well

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM, pls fix misnomers

**Pull request merged**
